src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,254 @@
+            ThrowExceptionIfCompleted();
+
+            if (Read())
+            {
+                switch (TokenType)
+                {
+                    case JsonTokenType.String:
+                        var value = GetString();
+
+                        return value.Split(DelimitedStringDelimiters, StringSplitOptions.RemoveEmptyEntries);
+
+                    default:
+                        var invalidCastException = new InvalidCastException();
+                        throw new JsonException(invalidCastException.Message, invalidCastException);
+                }
+            }
+
+            return null;
+        }
+
+        internal bool ReadNextTokenAsBoolOrFalse()
+        {
+            ThrowExceptionIfCompleted();
+
+            if (Read() && (TokenType == JsonTokenType.False || TokenType == JsonTokenType.True))
+            {
+                return GetBoolean();
+            }
+            return false;
+        }
+
+        internal IReadOnlyList<string> ReadNextStringOrArrayOfStringsAsReadOnlyList()
+        {
+            ThrowExceptionIfCompleted();
+
+            if (Read())
+            {
+                switch (_reader.TokenType)
+                {
+                    case JsonTokenType.String:
+                        return new[] { (string)_reader.GetString() };
+
+                    case JsonTokenType.StartArray:
+                        return ReadStringArrayAsReadOnlyListFromArrayStart();
+
+                    case JsonTokenType.StartObject:
+                        return null;
+                }
+            }
+
+            return null;
+        }
+
+        internal IReadOnlyList<string> ReadStringArrayAsReadOnlyListFromArrayStart()
+        {
+            ThrowExceptionIfCompleted();
+
+            List<string> strings = null;
+
+            while (Read() && _reader.TokenType != JsonTokenType.EndArray)
+            {
+                string value = _reader.ReadTokenAsString();
+
+                strings = strings ?? new List<string>();
+
+                strings.Add(value);
+            }
+
+            return (IReadOnlyList<string>)strings ?? Array.Empty<string>();
+        }
+
+        private void GetMoreBytesFromStream()
+        {
+            if (_reader.BytesConsumed < _buffer.Length)
+            {
+                ReadOnlySpan<byte> leftover = _buffer.AsSpan((int)_reader.BytesConsumed);
+                if (leftover.Length == _buffer.Length)
+                {
+                    ArrayPool<byte>.Shared.Return(_buffer);
+                    _buffer = ArrayPool<byte>.Shared.Rent(_buffer.Length * 2);
+                }
+                leftover.CopyTo(_buffer);
+                _stream.Read(_buffer, leftover.Length, _buffer.Length - leftover.Length);
+            }
+            else
+            {
+                _stream.Read(_buffer, 0, _buffer.Length);
+            }
+            _reader = new Utf8JsonReader(_buffer, isFinalBlock: _stream.Length == _stream.Position, _reader.CurrentState);
```

Comment:
I updated one of LockFIleFormatTests to have "hello" outside of the final close brace and it still passed. I also updated one of the tests in JsonSpecReaderTests and it also passes with "Hello" at the end of the JSON string. 

I tried it in dotnetfiddle because I couldn't find any documentation saying this was expected behavior. https://dotnetfiddle.net/snHuQm 

Anyway, that's weird. I'm not sure what you mean by capturing the results of _stream.Read, like keeping track of the number of bytes read from the stream and passing in only those specific bytes as a span to the Utf8JsonReader?

Reply:
Okay, Updated the way I'm populating the Utf8JsonReader so now only the data that was read form the stream is passed in, avoiding passing in a larger buffer with a bunch of empty items. 

If I pass in a malformed JSON to Utf8JsonReader, with text after the final closing bracket, it doesn't fail until the end if you do an additional read after the final close.

Reply:
> I tried it in dotnetfiddle because I couldn't find any documentation saying this was expected behavior

I find it very surprising that Newtonsoft.Json doesn't throw an exception. FWIW, if you try System.Text.Json's `JsonSerializer.Deserialize` method, it does validate that the end of the root object is the end of the file, and will throw if there's any trailing data. I really, really thought Newtonsoft.Json did the same!

> I'm not sure what you mean by capturing the results of _stream.Read, like keeping track of the number of bytes read from the stream and passing in only those specific bytes as a span to the Utf8JsonReader?

You've already updated your PR, but yes, that's exactly what I mean. If you read `Stream.Read`'s docs: https://learn.microsoft.com/en-us/dotnet/api/system.io.stream.read?view=net-8.0#system-io-stream-read(system-byte()-system-int32-system-int32)

>> An implementation is free to return fewer bytes than requested even if the end of the stream has not been reached.

So, there's no guarantee that the array is going to be filled, even when there is more data available in the stream. But in the case of reaching the end of the file, it's extremely unlikely that the last byte of the file will coincidentally match the last byte of the array.

Therefore, it's important to ensure that `Utf8JsonReader` isn't allowed to read past the end of the bytes that the `Stream` actually read. But it looks like you've fixed that now 👍